src/software/geom/algorithms/step_along_perimeter.cpp

```diff
@@ -0,0 +1,75 @@
+#include "software/geom/algorithms/step_along_perimeter.h"
+
+#include <algorithm>
+#include <vector>
+
+#include "software/geom/algorithms/collinear.h"
+#include "software/geom/algorithms/distance.h"
+#include "software/geom/segment.h"
+
+
+Point stepAlongPerimeter(const Polygon& polygon, const Point& start,
+                         double travel_distance)
+{
+    if (travel_distance == 0.0)
+    {
+        return start;
+    }
+
+    std::vector<Segment> polygon_segments = polygon.getSegments();
+    std::size_t start_segment_index       = 0;
+
+    // find initial segment which contains start point
+    auto it =
+        std::find_if(polygon_segments.begin(), polygon_segments.end(),
+                     [&start](const Segment& segment) {
+                         return collinear(segment.getStart(), start, segment.getEnd());
+                     });
+
+    start_segment_index = std::distance(polygon_segments.begin(), it);
+
+    std::size_t segment_index = start_segment_index;
+
+    // fmod travel distance for case where travel_distance > perimeter
+    bool is_counter_clockwise = travel_distance < 0;
+    travel_distance           = std::fmod(std::abs(travel_distance), polygon.perimeter());
+    if (is_counter_clockwise)
+    {
+        travel_distance = polygon.perimeter() - travel_distance;
```

Comment:
minor nit: we can store `perimeter` in a variable to avoid the possibility of double computation

Reply:
I assumed the compiler would probably take care of this but sure, will do

Reply:
The compiler might actually do the optimization, I'm not fully sure.